core/trino-spi/src/main/java/io/trino/spi/block/DictionaryBlock.java

```diff
@@ -445,14 +445,17 @@ public Block copyRegion(int position, int length)
     @Override
     public boolean mayHaveNull()
     {
-        return positionCount > 0 && dictionary.mayHaveNull();
+        return mayHaveNull;
     }
 
     @Override
     public boolean isNull(int position)
     {
+        if (!mayHaveNull) {
```

Comment:
Doing this before validation would appear to go against the contract defined in Block#isNull
```
 /**
     * Is the specified position null?
     *
     * @throws IllegalArgumentException if this position is not valid
     */
    boolean isNull(int position);
```

Reply:
I tried to figure out whether `@throws` in javadoc officially means a necessity to throw or that it may throw. No luck so far. Usually either `can` or `may` is used, but the context is more like "Programmer may put this into javadoc", rather than "Method can throw if ...".
The problem is that this change can actually make a difference. I've notice up to few % in some queries.
So the question is whether any code path actually relies on this. i.e. not throwing this exception here can cause trouble.

Reply:
I see 2 common cases here:
1. isNull is followed by getValue. This will immediately result in throwing the proper `IllegalArgumentException` from `getValue` method
2. More intenal usage when after `isNull` there is a direct access to the underlying storage. That would still result in exception, but rather a `ArrayIndexOutOfBoundException` or something alike

Reply:
In other blocks we have something like:
```
        checkReadablePosition(position);
        return valueIsNull != null && valueIsNull[position + arrayOffset];
```
so we check argument regardless of if nulls are possible or not.
I think we should be consistent between blocks in one way or another.

> The problem is that this change can actually make a difference.

`checkReadablePosition(position);` makes a difference. Shouldn't CPU make branch prediction?

Reply:
The more branches there are, the worse branch predictor will work. Also that may be a matter of code size and inlining.
These are all just wild guesses. The fact is that it does make a difference. And the question is what do want to do with it.
Making two methods `isNull` and `isNullUnsafe1 seems excessive.

Reply:
I'm fine keeping it. Although if this really makes a difference, we need to revisit other methods and blocks